### ISSUE TITLE
Restrict usage of special replacement patterns ($) in string string argument to `String.prototype.replace`

### ISSUE THREAD

**Issue by tom-sherman** [eslint/eslint#13521]
**Please describe what the rule should do:**

Restrict usage of special replacement pattern in string string argument to `String.prototype.replace`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#Specifying_a_string_as_a_parameter

These implicit replacement patterns are a foot gun more often than not, an eslint rule to ban them would be good.

**What new ECMAScript feature does this rule relate to?**

n/a

**What category of rule is this? (place an "X" next to just one item)**

[✔] Warns about a potential error (problem)
[✔] Suggests an alternate way of doing something (suggestion)
[] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

Bad:

```js
str.replace('a', 'foo $ bar')
str.replace('a', 'foo $$ bar')
str.replace('a', 'foo $` bar')
str.replace('a', "foo $' bar")
str.replace(/(a)(b)/g, '$1')
str.replace(/(a)(b)/g, 'foo $1 bar')
str.replace(/(?<name>a)(b)/g, 'foo $<name> bar')
```

All of these should be replaced with the function variant.

**Why should this rule be included in ESLint (instead of a plugin)?**

It's a core language feature that's esoteric enough to warrant wide adoption.

**Are you willing to submit a pull request to implement this rule?**

Yes, I've mainly opened this issue to remind me to open a PR at some point.

**Comment by mdjermanovic** [comment]
Hi @tom-sherman, thanks for the new rule proposal!

Per our new policy about rules (https://eslint.org/blog/2020/05/changes-to-rules-policies) we are no longer accepting new rules that are unrelated to new ECMAScript features.

Also, this looks to me more like a stylistic rule, since I'm not sure if the use of replacement functions instead of the `$` patterns is widely considered to be a best practice.

It's also possible to restrict this with the `no-restricted-syntax` rule:

```json
{
  "rules": {
    "no-restricted-syntax": [
      "error",
      {
        "selector": "CallExpression[callee.property.name='replace'] > Literal.arguments:nth-child(2)[value=/[$]/]",
        "message": "Use a function instead of replacement patterns."
      }
    ]  
  }
}
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1zeW50YXhcIjogW1xuICBcImVycm9yXCIsXG4gIHtcbiAgICBcInNlbGVjdG9yXCI6IFwiQ2FsbEV4cHJlc3Npb25bY2FsbGVlLnByb3BlcnR5Lm5hbWU9J3JlcGxhY2UnXSA+IExpdGVyYWwuYXJndW1lbnRzOm50aC1jaGlsZCgyKVt2YWx1ZT0vWyRdL11cIixcbiAgICBcIm1lc3NhZ2VcIjogXCJVc2UgYSBmdW5jdGlvbiBpbnN0ZWFkIG9mIHJlcGxhY2VtZW50IHBhdHRlcm5zLlwiXG4gIH1cbl0qL1xuXG5zdHIucmVwbGFjZSgnYScsICdmb28gJCBiYXInKVxuc3RyLnJlcGxhY2UoJ2EnLCAnZm9vICQkIGJhcicpXG5zdHIucmVwbGFjZSgnYScsICdmb28gJGAgYmFyJylcbnN0ci5yZXBsYWNlKCdhJywgXCJmb28gJCcgYmFyXCIpXG5zdHIucmVwbGFjZSgvKGEpKGIpL2csICckMScpXG5zdHIucmVwbGFjZSgvKGEpKGIpL2csICdmb28gJDEgYmFyJylcbnN0ci5yZXBsYWNlKC8oPzxuYW1lPmEpKGIpL2csICdmb28gJDxuYW1lPiBiYXInKSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

I'm leaving this issue open for the case that other team members have a different opinion.

**Comment by tom-sherman** [comment]
@mdjermanovic Thanks so much for this, I didn't know about the no-restricted-syntax rule - this should do me just fine!

> Per our new policy about rules (eslint.org/blog/2020/05/changes-to-rules-policies) we are no longer accepting new rules that are unrelated to new ECMAScript features.

I didn't know this, would probably help others a lot if this was added to the issue template. I'm happy for this to be closed

**Comment by mdjermanovic** [comment]
I'm glad it helped!

>  would probably help others a lot if this was added to the issue template

Thanks for the suggestion! We added `<!-- New rules must be related to ECMAScript features added within the last 12 months -->` comment to the new rule template, but maybe we should provide more context.
